pytorch_lightning/plugins/environments/bagua_environment.py

```diff
@@ -0,0 +1,65 @@
+# Copyright The PyTorch Lightning team.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import logging
+import os
+
+from pytorch_lightning.plugins.environments.cluster_environment import ClusterEnvironment
+
+log = logging.getLogger(__name__)
+
+
+class BaguaEnvironment(ClusterEnvironment):
```

Comment:
does bagua come with its own launcher or why do you need a separate cluster environment for it? Ideally we could handle it similar to how horovod does it.

I'm asking because a cluster (like slurm, kubeflow, redis etc.) usually is separated from the strategy. So if we can get those terms sorted out, it might be easier to use this in the end (ideally also with lightning creating the processes and setting the necessary environment variables when no external launcher is strictly required).

Reply:
Yes, Bagua has its own launcher, [here](https://tutorials.baguasys.com/getting-started/#launch-job) it the doc. Some optimizations in Bagua, like [Bagua-Net](https://tutorials.baguasys.com/more-optimizations/bagua-net) and [auto-tuning](https://tutorials.baguasys.com/performance-autotuning/), need to be enabled using Bagua launcher.

By design, Bagua is compatible with pytorch native distributed environment. We can launch a Bagua job using Pytorch launcher and start training using different [distributed algorithms](https://tutorials.baguasys.com/algorithms/) in Bagua, but the advanced optimizations mentioned above is off.

I have tested this strategy under Bagua env and Lightning env, but other env like slurm, kubeflow etc. has not tested yet ( It is not easy for me to have access to such envs).

And when using Bagua launcher, it can be automatically detected as a Bagua env. So that we have similar usage as horovod?

Reply:
I am not quite clear with how horovod handle this, it seems we can just launch the job using `horovodrun`.
As for bagua, we could do this similarly, by launching the job using `baguarun` or `bagua.distributed.launch`.

Reply:
I see, I was not sure, whether we need a environment plugin here, but since you also tested it with lightning env, that should be fine :)